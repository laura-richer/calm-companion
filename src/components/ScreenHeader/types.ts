import { ImageSourcePropType } from 'react-native';

export type ScreenHeaderProps = {
  title: string;
  image: ImageSourcePropType;
  imageWidth: number;
  imageHeight: number;
  imageFlushToBottom?: boolean;
};
