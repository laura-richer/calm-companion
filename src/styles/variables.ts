export const colors = {
  WHITE: '#ffffff',
  GREY: '#f7f7f7',
  ONE: '#3c433e',
  TWO: '#3f7b82',
  THREE: '#8bc3b9',
  FOUR: '#b4d4c0',
  FIVE: '#e1f1ce',
  DROP_SHADOW: '#8b8490',
};

export const spacer = {
  XS: 10,
  SM: 20,
  MD: 30,
  LG: 50,
};

export const fontFamily = {
  MEDIUM: 'Poppins-Medium',
  SEMIBOLD: 'Poppins-SemiBold',
  BOLD: 'Poppins-Bold',
};

export const fontSize = {
  XS: 8,
  SM: 10,
  MD: 14,
  LG: 18,
  XL: 36,
};

export const lineHeight = {
  default: 20,
};

export const borderRadius = {
  default: 8,
};
