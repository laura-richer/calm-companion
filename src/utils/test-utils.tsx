import React, { PropsWithChildren } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { render, RenderOptions } from '@testing-library/react-native';
import { Provider as StoreProvider } from 'react-redux';
import { EnhancedStore, configureStore } from '@reduxjs/toolkit';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { RootState } from '~store/index';
import navigationReducer, {
  initialState as navigationInitialState,
} from '~store/slices/navigation';
import quickHelpReducer, { initialState as quickHelpInitialState } from '~store/slices/quickHelp';

type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K];
};

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: EnhancedStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  { preloadedState = {}, store = createTestStore(preloadedState) }: ExtendedRenderOptions = {}
) => {
  const wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <SafeAreaProvider>
          <NavigationContainer>{children}</NavigationContainer>
        </SafeAreaProvider>
      </StoreProvider>
    );
  };

  return {
    ...render(ui, {
      wrapper,
    }),
  };
};

const _createStore = (preloadedState: Partial<RootState> = {}): EnhancedStore => {
  return configureStore({
    reducer: {
      navigation: navigationReducer,
      quickHelp: quickHelpReducer,
    },
    preloadedState,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        immutableCheck: { warnAfter: 256 },
        serializableCheck: { warnAfter: 256 },
      }),
  });
};

export const createTestStore = (preloadedState: DeepPartial<RootState> = {}): EnhancedStore => {
  const state: RootState = {
    // @ts-ignore
    navigation: {
      ...navigationInitialState,
      ...(preloadedState.navigation || {}),
    },
    // @ts-ignore
    quickHelp: {
      ...quickHelpInitialState,
      ...(preloadedState.quickHelp || {}),
    },
  };
  const store: EnhancedStore = _createStore(state);
  const origDispatch = store.dispatch;
  store.dispatch = jest.fn(origDispatch);
  return store;
};
