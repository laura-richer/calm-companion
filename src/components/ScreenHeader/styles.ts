import { StyleSheet } from 'react-native';
import { colors, spacer } from '~styles/variables';

export default StyleSheet.create({
  container: {
    padding: spacer.XS,
  },
  containerFlushToBottom: {
    paddingBottom: 0,
  },
  title: {
    color: colors.WHITE,
    marginBottom: spacer.XS,
  },
  imageContainer: {
    alignItems: 'flex-end',
  },
  image: {
    opacity: 0.1,
  },
  imageFlushToBottom: {
    marginTop: spacer.XS,
  },
});
