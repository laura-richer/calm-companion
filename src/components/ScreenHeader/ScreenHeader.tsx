import React from 'react';
import { View, Text, Image } from 'react-native';
import GradientDark from '~components/Gradient/GradientDark';
import { ScreenHeaderProps } from './types';
import { typographyStyles } from '~styles/global';
import styles from './styles';

const ScreenHeader = ({
  title,
  image,
  imageWidth,
  imageHeight,
  imageFlushToBottom = false,
}: ScreenHeaderProps) => {
  return (
    <GradientDark>
      <View
        testID="screenHeaderContainer"
        style={[styles.container, imageFlushToBottom ? styles.containerFlushToBottom : null]}
      >
        <Text style={[typographyStyles.textLg, styles.title]}>{title}</Text>
        <View style={styles.imageContainer}>
          <Image
            accessibilityLabel={`${title} image`}
            style={[styles.image, imageFlushToBottom ? styles.imageFlushToBottom : null]}
            source={image}
            resizeMode="contain"
            width={imageWidth}
            height={imageHeight}
          />
        </View>
      </View>
    </GradientDark>
  );
};

export default ScreenHeader;
