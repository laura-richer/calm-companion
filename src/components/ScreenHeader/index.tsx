import React from 'react';
import { View, Text, Image } from 'react-native';
import GradientDark from '~components/Gradient/GradientDark';
import { ScreenHeaderProps } from './types';
import { typographyStyles } from '~styles/global';
import styles from './styles';

const ScreenHeader = ({ title, icon }: ScreenHeaderProps) => {
  return (
    <GradientDark>
      <View style={styles.container}>
        <Text style={[typographyStyles.titleLg, styles.title]}>{title}</Text>
      </View>
      {/* <Image source={require(icon)} /> */}
    </GradientDark>
  );
};

export default ScreenHeader;
