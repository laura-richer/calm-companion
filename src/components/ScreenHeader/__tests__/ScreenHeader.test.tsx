import React from 'react';
import { screen } from '@testing-library/react-native';
import { renderWithProviders } from '~utils/test-utils';
import ScreenHeader from '../ScreenHeader';
import { spacer } from '~styles/variables';

const _renderDefault = () => {
  return renderWithProviders(
    <ScreenHeader
      title={'Quick Help'}
      image={require('~assets/images/screenHeaders/quick-help.png')}
      imageWidth={117}
      imageHeight={75}
    />
  );
};

describe('screen header', () => {
  beforeEach(() => {
    _renderDefault();
  });

  test('matches snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('renders title', () => {
    const title = screen.getByText('Quick Help');
    expect(title).toBeOnTheScreen();
  });

  test('renders correct image', () => {
    const image = screen.getByLabelText('Quick Help image');
    expect(image).toBeOnTheScreen();
    expect(image).toHaveProp('source', require('~assets/images/screenHeaders/quick-help.png'));
  });

  test('renders image with correct width', () => {
    const image = screen.getByLabelText('Quick Help image');
    expect(image).toHaveProp('width', 117);
  });

  test('renders image with correct height', () => {
    const image = screen.getByLabelText('Quick Help image');
    expect(image).toHaveProp('height', 75);
  });

  test('renders image with correct resizeMode', () => {
    const image = screen.getByLabelText('Quick Help image');
    expect(image).toHaveProp('resizeMode', 'contain');
  });

  test('renders image with default margin', () => {
    const screenHeaderContainer = screen.getByTestId('screenHeaderContainer');
    expect(screenHeaderContainer).toHaveStyle({ padding: spacer.XS });
  });

  test('renders image flush to bottom', () => {
    renderWithProviders(
      <ScreenHeader
        title={'Quick Help'}
        image={require('~assets/images/screenHeaders/quick-help.png')}
        imageWidth={117}
        imageHeight={75}
        imageFlushToBottom={true}
      />
    );

    const image = screen.getByTestId('screenHeaderContainer');
    expect(image).toHaveStyle({ paddingBottom: 0 });
  });
});
