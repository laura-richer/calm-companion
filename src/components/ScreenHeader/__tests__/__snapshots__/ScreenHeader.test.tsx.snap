// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screen header matches snapshot 1`] = `
<BVLinearGradient
  angle={170}
  angleCenter={
    {
      "x": 0.75,
      "y": 0.75,
    }
  }
  colors={
    [
      4282350466,
      4287349689,
    ]
  }
  endPoint={
    {
      "x": 0.5,
      "y": 1,
    }
  }
  locations={null}
  startPoint={
    {
      "x": 0.5,
      "y": 0,
    }
  }
  useAngle={true}
>
  <View
    style={
      [
        {
          "padding": 10,
        },
        null,
      ]
    }
    testID="screenHeaderContainer"
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Bold",
            "fontSize": 36,
          },
          {
            "color": "#ffffff",
            "marginBottom": 10,
          },
        ]
      }
    >
      Quick Help
    </Text>
    <View
      style={
        {
          "alignItems": "flex-end",
        }
      }
    >
      <Image
        accessibilityLabel="Quick Help image"
        height={75}
        resizeMode="contain"
        source={
          {
            "testUri": "../../../src/assets/images/screenHeaders/quick-help.png",
          }
        }
        style={
          [
            {
              "opacity": 0.1,
            },
            null,
          ]
        }
        width={117}
      />
    </View>
  </View>
</BVLinearGradient>
`;
